### DISCUSSION TITLE
Emit a named `val` in DSL2

### DISCUSSION THREAD

**Discussion by multimeric** [nextflow-io/nextflow#2115] [Q&A] [answered]
I like to name all my outputs. However, if you are outputting a simple value, it is seemingly impossible to assign a name to it in DSL2.

If I try to use `emit`:
```groovy
nextflow.enable.dsl=2

process test {
    output:
        val "abc", emit x
    script:
        """
        true
        """
}

workflow {
    result = test()
    result.x.view()
}
```
```
No such variable: emit
```

If I try to re-assign a new variable:
```groovy
nextflow.enable.dsl=2

process test {
    output:
        val x
    script:
        x = "abc"
        """
        true
        """
}

workflow {
    result = test()
    result.x.view()
}
```
```
Missing process or function with name 'view'
```
(because this isn't a named input, so only `result.view()` would work)

And of course I can't use `into`:
```groovy
nextflow.enable.dsl=2

process test {
    output:
        val "abc" into x
    script:
        """
        true
        """
}

workflow {
    result = test()
    result.x.view()
}
```
```
Process clause `into` should not be provided when using DSL 2
```

How can this be done?

**Comment by phue** [comment, marked as answer]
The syntax requires to put a colon after the `emit` keyword, i.e:

```
nextflow.enable.dsl=2

process test {
    output:
        val "abc", emit: x
    script:
        """
        true
        """
}

workflow {
    test()
    test.out.x.view()
}

```

**Reply by multimeric** [reply]
Good point. I keep making syntax errors and thinking they are a deeper issue.

**Reply by phue** [reply]
Now that's odd, the above snippet works for me with nextflow `21.04.0`

**Reply by multimeric** [reply]
Ah no you are right, that only happened when I forgot to enable DSL2.

**Reply by bobamess** [reply]
I see you've also corrected the other error.  The original code had `test.x.view()`, which is why it continued to fail, but now has the correct `test.out.x.view()` but this was not commented on.

**Reply by multimeric** [reply]
No, my original code had `result.x.view()`, which is equivalent to `test.out.x.view()`.

**Reply by bobamess** [reply]
Sorry. You're right, of course, so it did. @phue made the change to `test.out.x.view()` without commenting and I overlooked the `result = test()` bit in your original.
